portable/GCC/ARM_CRx_MPU/port.c

```diff
@@ -0,0 +1,799 @@
+    /* Privileged Write and Read, Unprivileged Read, MPU Region for PRIVILEGED_DATA. */
+    ulRegionStart = ( uint32_t ) __privileged_data_start__;
+    ulRegionEnd = ( uint32_t ) __privileged_data_end__;
+    ulRegionLength = ulRegionEnd - ulRegionStart;
+    ulRegionLength = prvGetMPURegionSizeSetting( ulRegionLength ) | portMPU_REGION_ENABLE;
+    prvMpuSetRegion(
+        portPRIVILEGED_RAM_REGION,
+        ulRegionStart,
+        ulRegionLength,
+        portMPU_PRIV_RW_USER_RO_NOEXEC | portMPU_NORMAL_OIWTNOWA_SHARED
+    );
+
+    /* After setting default regions, enable the MPU */
+    prvMpuEnable();
+}
+
+/* ------------------------------------------------------------------------- */
+
+/** @brief Determine if a FreeRTOS Task has been granted access to a memory region
+ *
+ * @param xTaskMPURegion Pointer to a single set of MPU region registers.
+ * @param ulRegionStart Base address of the memory region access is being requested.
+ * @param ulRegionLength The length of the memory region that access is being requested.
+ * @param ulAccessRequested The type of access being requested, either read or write.
+ * @return BaseType_t pdTRUE if the task can access the region, pdFALSE otherwise
+ *
+ * @ingroup Task Context
+ * @ingroup MPU Control
+ */
+PRIVILEGED_FUNCTION static BaseType_t prvTaskCanAccessRegion(
+    const xMPU_REGION_REGISTERS * xTaskMPURegion,
+    const uint32_t ulRegionStart,
+    const uint32_t ulRegionLength,
+    const uint32_t ulAccessRequested
+)
+{
+    BaseType_t xAccessGranted;
+    uint32_t ulRegionEnd = ulRegionStart + ulRegionLength;
+
+    /* Convert the MPU Region Size value to the actual size */
+    uint32_t ulTaskRegionLength = 1 << ( ( xTaskMPURegion->ulRegionSize >> 1 ) + 1U );
+    // uint32_t ulTaskRegionLength = 2 << ( xTaskMPURegion->ulRegionSize >> 1 );
+    uint32_t ulTaskRegionEnd = xTaskMPURegion->ulRegionBaseAddress + ulTaskRegionLength;
+    if( ( ulRegionStart >= xTaskMPURegion->ulRegionBaseAddress ) &&
+        ( ulRegionEnd <= ulTaskRegionEnd ) )
+    {
+        /* Unprivileged read is MPU Ctrl Access Bit Value bX1X */
+        if( ( tskMPU_READ_PERMISSION == ulAccessRequested ) &&
+            ( ( portMPU_PRIV_RW_USER_RO_NOEXEC ) &xTaskMPURegion->ulRegionAttribute ) )
+        {
+            xAccessGranted = pdTRUE;
+        }
+
+        /* Unprivileged Write is MPU Ctrl Access Bit Value b011 */
+        else if( ( tskMPU_WRITE_PERMISSION & ulAccessRequested ) &&
+                ( portMPU_PRIV_RW_USER_RW_NOEXEC ==
+                ( portMPU_PRIV_RW_USER_RW_NOEXEC & xTaskMPURegion->ulRegionAttribute ) ) )
+        {
+            xAccessGranted = pdTRUE;
+        }
+
+        else
+        {
+            xAccessGranted = pdFALSE;
+        }
+    }
+    else
+    {
+        xAccessGranted = pdFALSE;
+    }
+
+    return xAccessGranted;
+}
+
+/* ------------------------------------------------------------------------- */
+
+/* PRIVILEGED_FUNCTION */ BaseType_t xPortIsAuthorizedToAccessBuffer(
```

Comment:
There are several functions with the  PRIVILEGED_FUNCTION decorator commented out. Should they in fact be in the privileged region? If not, can this comment be deleted

Reply:
This comment is here to make it clear that the function is in privileged functions.

The declarations have the attributes that place them into PRIVILEGED_FUNCTIONS. This is the standard used in this repository.